es/es-repository-s3/src/test/java/org/elasticsearch/repositories/s3/S3BlobStoreContainerTests.java

```diff
@@ -0,0 +1,409 @@
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package org.elasticsearch.repositories.s3;
+
+import com.amazonaws.AmazonClientException;
+import com.amazonaws.services.s3.AmazonS3;
+import com.amazonaws.services.s3.model.AbortMultipartUploadRequest;
+import com.amazonaws.services.s3.model.CannedAccessControlList;
+import com.amazonaws.services.s3.model.CompleteMultipartUploadRequest;
+import com.amazonaws.services.s3.model.CompleteMultipartUploadResult;
+import com.amazonaws.services.s3.model.InitiateMultipartUploadRequest;
+import com.amazonaws.services.s3.model.InitiateMultipartUploadResult;
+import com.amazonaws.services.s3.model.ObjectMetadata;
+import com.amazonaws.services.s3.model.PartETag;
+import com.amazonaws.services.s3.model.PutObjectRequest;
+import com.amazonaws.services.s3.model.PutObjectResult;
+import com.amazonaws.services.s3.model.StorageClass;
+import com.amazonaws.services.s3.model.UploadPartRequest;
+import com.amazonaws.services.s3.model.UploadPartResult;
+import org.elasticsearch.common.blobstore.BlobPath;
+import org.elasticsearch.common.blobstore.BlobStore;
+import org.elasticsearch.common.collect.Tuple;
+import org.elasticsearch.common.unit.ByteSizeUnit;
+import org.elasticsearch.repositories.ESBlobStoreContainerTestCase;
+import org.mockito.ArgumentCaptor;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import static org.elasticsearch.repositories.s3.S3BlobStoreTests.randomMockS3BlobStore;
+import static org.hamcrest.Matchers.instanceOf;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.doAnswer;
+import static org.mockito.Mockito.doNothing;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+public class S3BlobStoreContainerTests extends ESBlobStoreContainerTestCase {
+
+    protected BlobStore newBlobStore() {
+        return randomMockS3BlobStore();
+    }
+
+    @Override
+    public void testDeleteBlob() {
+        assumeFalse("not implemented because of S3's weak consistency model", true);
+    }
+
+    @Override
+    public void testVerifyOverwriteFails() {
+        assumeFalse("not implemented because of S3's weak consistency model", true);
+    }
+
+    public void testExecuteSingleUploadBlobSizeTooLarge() {
+        final long blobSize = ByteSizeUnit.GB.toBytes(randomIntBetween(6, 10));
+        final S3BlobStore blobStore = mock(S3BlobStore.class);
+        final S3BlobContainer blobContainer = new S3BlobContainer(mock(BlobPath.class), blobStore);
+
+        final IllegalArgumentException e = expectThrows(IllegalArgumentException.class, () ->
```

Comment:
shall we use rules?

Reply:
I'd prefer it as it increases readability imo

Reply:
Keep in mind that the `ExpectedException` aborts the test. E.g. the following will pass:

```
    @Test
    public void testFoo() {
        expectedException.expectMessage("foo");
        if (true) {
            throw new IllegalArgumentException("foo");
        }
        assertThat(1, is(2));
    }
```

Due to that I'm wondering if we should go the other route, and start to favour `expectThrows` over the rule.

Reply:
I wonder if we ever had a problem with that..

Reply:
I think I've seen it once or twice over the last couple of years, so it is not a big issue. Just something to be aware off. (And when I ported over other ES tests I would have had to make more substantial changes, because the tests were written in a way that they had to continue after the exception checks)

Reply:
I see. I also have no strong opinion here just a preference.

Reply:
Doesn't have to be black or white. IMHO, we could use the rule whenever possible (as I also agree that it favors readability) and stick with `expectThrows` when the test needs to continue with further assertions. Just my 2c

Reply:
pushed the fixup. used rules where it makes